### DISCUSSION TITLE
Textadept IDE and PML

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#59] [Editors Support]
The [Textadept] IDE might provide an interesting solution for PML.

It's a FOSS and cross platform editor which is very nice, feature rich, and fully customizable, which makes it a good candidate for creating an _ad hoc_ edition for PML — i.e. a pre-packaged version with all the tweaks and plugins for working with PML. Most important, it's _very well_ [documented].

Under Windows, it's also available as a standalone version (unZip and use), which makes it very easy to customize it without having to deal with package managers, extra tool, etc.

Although its core is written in C, it's a Lua powered editor, so all extensions are written in Lua.

LSP is natively supported via a Lua module:

https://github.com/orbitalquark/textadept-lsp

which means it should be fairly easy to write a PML Lang Server, with Lua being a friendly scripting language, and the whole Textadept setup being very easy to handle (you basically work just inside the application directory tree, no settings and packages scattered all over the place).

Of course, once you have a PML Lang Server in Lua it can then be used on other editors too, you only need to bundle the required Lua runtime in the various packages (which is small) and write a basic package for the target editor, to interface it.

[Textadept] is a really cool editor, small sized, fully portable (works even on a USB stick) and has very small memory requirements. It's a Scintilla based editor, but comes with all the modern features like multi-cursor editing, fuzzy matching, etc. If I ever needed to create a pre-packaged editor for the needs of a specific domain, [Textadept] would be my first choice editor for the task — especially since updating it would be just a matter of dropping the updated files from the original package into the tweaked version, no re-compiling needed.

- https://github.com/orbitalquark/textadept
- https://github.com/orbitalquark/textadept/wiki

[Textadept]: https://orbitalquark.github.io/textadept/
[documented]: https://orbitalquark.github.io/textadept/api.html

**Comment by celtic-coder** [comment]
Hi Tristano (@tajmone),

I've looked into TextAdept as well when I was investigating coding in Lua. It looks quite an accomplished editor.

I'll be looking further into it when we have our house refurbishment completed and we can get it setup for sale.

Kind Regards,
Liam
